src/libsyntax/parse/diagnostics.rs

```diff
@@ -561,17 +571,116 @@ impl<'a> Parser<'a> {
+                                // `ExprKind::Err` placeholder.
+                                return Ok(Some(self.mk_expr(
+                                    lhs.span.to(self.prev_span),
+                                    ExprKind::Err,
+                                    ThinVec::new(),
+                                )));
+                            }
+                            Err(mut err) => {
+                                err.cancel();
+                                // Not entirely sure now, but we bubble the error up with the
+                                // suggestion.
+                                mem::replace(self, snapshot);
+                                return Err(err);
+                            }
+                        }
+                    } else if token::OpenDelim(token::Paren) == self.token.kind {
+                        // We have high certainty that this was a bad turbofish at this point.
+                        // `foo< bar >(`
+                        err.span_suggestion(
+                            op_span.shrink_to_lo(),
+                            msg,
+                            "::".to_string(),
+                            Applicability::MaybeIncorrect,
+                        );
+
+                        let snapshot = self.clone();
+                        self.bump(); // `(`
+
+                        // Consume the fn call arguments.
+                        let modifiers = vec![
+                            (token::OpenDelim(token::Paren), 1),
+                            (token::CloseDelim(token::Paren), -1),
+                        ];
+                        let early_return = vec![token::Eof];
+                        self.consume_tts(1, &modifiers[..], &early_return[..]);
+
+                        if self.token.kind == token::Eof {
+                            // Not entirely sure now, but we bubble the error up with the
+                            // suggestion.
+                            mem::replace(self, snapshot);
+                            return Err(err);
+                        } else {
+                            // 99% certain that the suggestion is correct, continue parsing.
+                            err.emit();
+                            // FIXME: actually check that the two expressions in the binop are
+                            // paths and resynthesize new fn call expression instead of using
+                            // `ExprKind::Err` placeholder.
+                            return Ok(Some(self.mk_expr(
+                                lhs.span.to(self.prev_span),
+                                ExprKind::Err,
+                                ThinVec::new(),
+                            )));
+                        }
+                    } else {
+                        // All we know is that this is `foo < bar >` and *nothing* else. Try to
+                        // be helpful, but don't attempt to recover.
```

Comment:
So if we have e.g. `identity<u8>` we don't get the suggestion here which is a shame... maybe change that?

Also I think this branch isn't exercised by a test anymore.

Reply:
Added a test, let me know what you think.